### PR TITLE
docker/-compose refactoring + start.py script with docs

### LINKED ISSUE #98: Divide requirements.txt into development.txt and production.txt
We could separate the `requirements.txt` file into small files such as development/local.txt and requirements/production.txt.

_Originally posted by @Andrew-Chen-Wang in https://github.com/intelowlproject/IntelOwl/issues/95#issuecomment-647153922_

### LINKED ISSUE #277: Divide common docker services into diff. compose files
>  What I meant is basically your idea but applied to all the services, not only the ones related to celery.

Ah, I see. Yes, definitely what I meant as well. So, let's create a `compose.common.yml` file with the services like `rabbit-mq`, `postgres` which are common in all compose files and similarly for celery.

_Originally posted by @Eshaan7 in https://github.com/intelowlproject/IntelOwl/pull/274#issuecomment-737187816_

### PR DESCRIPTION
**Pull request by 0ssigeno**
# Description
Moved docker files in a subdirectory
Created start.py
Fixed docker integrations files
Added documentation

## Fixes
extended from #279 
closes #277
closes #98
## Type of change

- [x] Bug fix (non-breaking change which fixes an issue).
- [x] New feature (non-breaking change which adds functionality).
- [x] Breaking change (fix or feature that would cause existing functionality to not work as expected).

# Checklist

- [x] The pull request is for the branch develop
- [x] I changed the file [Usage](https://github.com/intelowlproject/IntelOwl/blob/master/docs/source/Usage.md) and [ReadMe](https://github.com/intelowlproject/IntelOwl/blob/master/README.md) if I added a new analyzer. 
- [x] The tests gave 0 errors.
- [x] `Black` gave 0 errors.
- [x] `Flake` gave 0 errors.
- [ ] I squashed the commits into a single one.

### CODE REVIEW

File: docs/source/Tests.md [hunk 1]
```diff
@@ -19,7 +19,7 @@
 ### Launch tests
     
 - Start the app using the docker-compose test file. In this way, you would launch the code in your environment and not the last official image in Docker Hub:
-`docker-compose -f docker-compose-for-tests.yml up`
+`docker-compose -f docker/default.yml -f docker/test.yml up`
```
Comment (Eshaan7) [inline docs/source/Tests.md:22]: correction: should be: `docker/test.override.yml`

File: start.py [hunk 2]
```diff
@@ -0,0 +1,92 @@
...
+    context_settings=dict(
+        ignore_unknown_options=True,
+        allow_extra_args=True,
+    )
+)
+@click.argument("mode", required=True, type=click.Choice(["prod", "test", "ci"]))
+@click.argument(
+    "docker_command", required=True, type=click.Choice(["build", "up", "down"])
+)
+@click.option("--traefik", default=False, is_flag=True)
```
Comment (Eshaan7) [inline start.py:39]: really think we should provide some help with each option. For example, `help="Uses the `traefik.override.yml` compose file"` and similarly for other options.

**Comment by mlodic** [comment]
I think we should merge this after having pushed #282, #290 and #283. New PRs will need to be based from this refactor

File: requirements.txt [hunk 3]
```diff
@@ -67,8 +67,6 @@ quark-engine==20.12
 git+https://github.com/DissectMalware/XLMMacroDeobfuscator.git@89fbce0c87014a4b5a22c1aef09c8b3ea9bf16c0
 speakeasy-emulator==1.4.8
 sphinx_rtd_theme==0.5.0
-flake8==3.8.4
-black==20.8b1
 pre-commit==2.9.2
```
Comment (Eshaan7) [inline requirements.txt:70]: sphinx_rtd_theme and pre-commit are also part of testing/development

File: docker/docker_entrypoint.sh [hunk 4]
```diff
@@ -6,4 +6,6 @@
 python manage.py migrate                  # Apply database migrations
 python manage.py collectstatic --noinput  # Collect static files
 
+bash api_app/script_analyzers/yara_repo_downloader.sh
```
Comment (Eshaan7) [inline docker/docker_entrypoint.sh:9]: should revert this change. The yara/quark-engine rules are to be kept constant for a particular IntelOwl release (otherwise, there are bugs). If we make this script part of entrypoint, then it gets executed each time the container is started thus slowing it down by a lot.

Comment (mlodic) [reply docker/docker_entrypoint.sh:9]: that's right... the idea was to reduce the size of the base image + reduce the build time of the image.

I would like to avoid to load the rules in any case at the start. I think they should be loaded only the first time they are requested and only if the users want to use them. Another motivation: the number of Yara repositories is growing a lot so I do not think they should be managed in the base image. It does not scale. To solve this we would need an additional work that is maybe out of the scope of the PR.

We could just create an issue for this and revert this change. Thoughts?

Comment (Eshaan7) [reply docker/docker_entrypoint.sh:9]: > For now, we could just create an issue for this and revert this change.

I agree.

> To solve this we would need an additional work that is maybe out of the scope of the PR.

Yeah

File: requirements.txt (hunk 3 above)
Comment (0ssigeno) [reply requirements.txt:70]: You don't have a `.readthedocs.yml` file. 
If i move the sphinx import, the readthedocs should not build anymore.

**Comment by Eshaan7** [comment]
Finally, I think everything is satisfactory. :D
Very nice work, @0ssigeno. Thank you for your patience through every iteration. 

Before we merge this PR, there's still 1 change left. After we have merged the other open PRs, we would need to modify the `docker/env_file_app_template` and `docker/env_file_app_ci`  to add entries for the newly added variables and cross-check that no variable is left.

**Comment by 0ssigeno** [comment]
I did not found any problem during the merge, so I think everything is fine. If you have time, you can doublecheck the environ variables, just to be sure.
If problems for readthedocs arise after the merge, I will try to fix them asap.

File: .readthedocs.yml [hunk 5]
```diff
@@ -0,0 +1,24 @@
...
+  image: latest
+
+# Build all formats
+formats: all
+
+# Optionally set the version of Python and requirements required to build your docs
+python:
+  version: 3.7
+  install:
+    - requirements: docs/requirements.txt
```
Comment (Eshaan7) [inline .readthedocs.yml:18]: this is set to `docs/requirements.txt`

File: docs/source/requirements.txt [hunk 6]
```diff
@@ -1 +1 @@
-# This requirements.txt is intentionally left blank for RTD build
```
Comment (Eshaan7) [inline docs/source/requirements.txt:1]: the path here is `docs/source/requirements.txt`. I'd prefer to move this file to `docs/requirements.txt`.
